### ISSUE TITLE
Arbitrarily responds in Spanish

### ISSUE THREAD

**Issue by axamaxa** [logancyang/obsidian-copilot#927]
- [x ] Disable all other plugins besides Copilot **(required)**
- [x ] Screenshot of note + Copilot chat pane + dev console added **(required)**
<img width="710" alt="Arbitrary Spanish response 1" src="https://github.com/user-attachments/assets/c23d058d-160b-441f-8f1c-7654f38f30c3" />

As you can see, the response is in Spanish and the sources are all in English.
<img width="710" alt="Arbitrary Spanish response 1" src="https://github.com/user-attachments/assets/90c6d00b-aade-47bd-9a36-6380e7323999" />

Copilot version: 2.7.9

(Bug report without the above will be closed)

**Describe how to reproduce**
A clear and concise description of what the bug is. Clear steps to reproduce the behavior
After several responses in English (native language), suddenly responds ad hoc in Spanish for one answer, then returns to English for several answers, then arbitrarily answers in Spanish again. Sometimes get stuck in Spanish. If I ask it to translate into English it might. But other times it might give a different answer in Spanish. Sometimes it might summarize in Spanish when asked to translate answer into English. Same if I highlight in focus panel and invoke the translate to english function: sometimes it does, sometimes it makes up something else in Spanish. Occasionally, it breaks altogether and responds with a super-long answer in unintelligible english, or a long quote from some unrelated note.
**Expected behavior**
Due to proper functioning until the psychotic break occurs, I have come to expect answers in English (native tongue and what the notes are written in .....no notes are written in Spanish)
**Screenshots**
Add screenshots to help explain your problem. Please turn on debug mode in Copilot settings, turn off other plugins to leave only Copilot dev messages as necessary.

**Additional context**
The first several responses/answers are usually very high quality and in English. At some arbitrary point in the discussion, it breaks open and starts speaking Spanish. Sometimes (rarely) I can bring it back for awhile, but sometimes ....he's just gone.

**Comment by logancyang** [comment]
what model are you using? Can please show the dev console too.

**Comment by axamaxa** [comment]
Now it's responding in French ....arbitrarily, sometimes in English. I've been able to employ the co-pilot translate function successfully (from the '/' command menu) and it is translating wonderfully tonight. But this costs me double in terms of my API (once for the French and once again for the translation). I appreciate the polyglottal excellence of the ai but the UE is not very efficient :-). I'm using the open.ai 4O model. I'll send you a screen-capture of the dev console too, but need clarification on what that is and exactly which page of it. I pulled up a bunch of (what looked like HTML code, or something) and tried to read it, because the instructions said to ...but I couldn't find anything although I'm not very good at that. I also clicked a tab that showed (what looked like an 'error log'). Is that what you want me to send? Just forebear my inexpertise and describe the 'dev console' item (where to find it and what part you want from it) and I'll get it to you. Bug report is turned on. Thanks .....before this started happening, the plug-in was awesome and a real game-changer for the Obsidian archives!

**Comment by logancyang** [comment]
The main thing I'm looking for is the context it retrieved from your notes, that's the first suspect. If it retrieved some random stuff like, e.g. excalidraw diagrams, the model might be tricked to speak in other languages. 

You can check the dev console with debug mode on, there should be info related to retrieval.

**Comment by axamaxa** [comment]
now it's decided to speak Spanish and it won't translate from the co-pilot 'translate to english' command. It just makes up more answers in Spanish. I believe this behavior emerges from the co-pilot settings in the section where we enter (something to the effect of:) 'how many previous questions and responses should co-pilot look at, in consideration of the next answer'. In other words, once the non-native responses get started, a feedback loop is generated which favors more non-native languages. I will go back into settings and turn this down to zero, if I can (or the lowest value allowed). Let's say this emends the on-going 'foreign response' cascade, the question then becomes 'what sets this feedback loop off in the first place?' and why does it go for 'x' number of english responses before arbitrarily deciding to answer in other languages?

I have excalidraw on the machine but it is turned off, and there are no notes in my vault with excalidraw. All my notes (to the best of my knowledge) are entirely in english. Some notes do, however, have phrases from classical Latin (which might look like Spanish to the ai, but I doubt it ...ai is great at pegging languages and likely wouldn't be fooled. There are a few notes that have words that are coined from ancient Greek terms (and with that, a scattering of Greek alphabet). But these notes are largely in english and, if co-pilot is so sensitive that there is no tolerance for a greek word or two (for example 'agape' love), then it will be mostly useless for Christians, Muslims, Jews, Scholars, and bi-lingual emails, etc. etc. Trust me, the foreign words scattered around my vault are a miniscule fraction compared to the 35 million english words. It occurs to me that there is no selection for 'preferred language' in the copilot settings. Perhaps co-pilot inherits this preference from the Obsidian preferences?

I deleted the 'wrong language' responses (desperate people do desperate things in hopes that something might work) because I was afraid they would influence the ai as it referenced our backlogs. So I will need to save these for you as it happens again ...it's a fair request. Especially the queries (although they are all english and consist of: 'define and discuss the concept of 'x', according to the information in my notes'. That never varies. The odd language thing just emerges arbitrarily after many successful responses sometime in the middle of a session. And I'll go back into the dev console and try to get something for you from there too. It may not be tonight (I'm whupped) but as soon as I can. Thanks! ......you know what, this reminds me: everything was going well (all english responses) until I asked the ai about metaphysics, and then I got the French answer. And I had a fleeting hunch at the time that the word 'metaphysics' may have sparked a flurry of Italian answers in the past. I don't know this for sure, but I remember having a strong hunch. So now, tired and sleepy, I've gone back into earlier 'saved chats' and pulled up that response and made a screenshot of the query. The following is the text part: 

"**user**: please define and discuss 'metaphysics', really emphasizing the concepts and metaphysical connotations present in our Obsidian notes.
[Timestamp: 2024/12/13 17:53:32]

**ai**: La **métaphysique** est une branche de la philosophie qui s'intéresse à l'étude de la nature fondamentale de l'être et du monde. Dans nos notes Obsidian, il est mentionné que la métaphysique tente de répondre à des questions essentielles telles que "Qu'est-ce qui existe ?" et "Quel est son caractère ?". Les métaphysiciens cherchent à clarifier les notions fondamentales qui nous permettent de comprendre le monde, comme l'existence, les objets et leurs  "... and it continues in french. I'll include a screenshot of this in an attachment. Now to bed for a while. I'll get the other stuff to you tomorrow. Thanks.

<img width="718" alt="French response from co-pilot" src="https://github.com/user-attachments/assets/fce017e6-e557-4134-ad1f-fcc742774b15" />

<img width="718" alt="French response from co-pilot" src="https://github.com/user-attachments/assets/decb5a40-b02b-4616-8a1c-88b2a6b7023b" />

<img width="718" alt="French response from co-pilot" src="https://github.com/user-attachments/assets/99dd7dc5-6f86-4064-8b54-cd6d24ae3d55" />

**Comment by logancyang** [comment]
Can you try add this sentence to your system prompt in the copilot setting: "Always respond in English." and see if it makes a difference?

**Comment by axamaxa** [comment]
It was speaking in Spanish or Italian today, right from jump street and I was able to translate it with the 'command 'P' function, but I was a little despondent since it started right in on this at the beginning. Then I found your reply and added the instructions you sent in the custom prompts and it cleared right up. I hope this fixes it for good (I'll keep you posted). This was a brilliant recommendation (I should have thought of it myself, though). Kudos to you: simple but effective!!! Thanks a million!
